vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java

```diff
@@ -4663,7 +4663,8 @@ private void onInMemoryFilterOrSortingChange(
         updateInMemorySorting(sortComparator);
         updateInMemoryFiltering(filter);
 
-        dataCommunicator.reset();
+        // Call refreshAll as data may have changed due filter change
+        getDataProvider().refreshAll();
```

Comment:
I propose to move `refreshAll()` from here to the data view.
Then the `setFilter` will be like on the Flow side:
```
    @Override
    public AbstractListDataView<T> setFilter(SerializablePredicate<T> filter) {
        DataViewUtils.setComponentFilter(component, filter);
        fireFilteringOrSortingChangeEvent(filter,
                (SerializableComparator<T>) DataViewUtils
                        .getComponentSortComparator(component).orElse(null));
        refreshAll();
        return this;
    }
```
This would fix all similar issues in data view, not only the Grid.